File: Leetcode/Algorithms/Easy/DistributeCandies/DistributeCandies.cpp
// Method - 1
class Solution {
public:
    int distributeCandies(vector<int>& candyType) {
        int unique_element = 0;
        
        sort(candyType.begin(), candyType.end());
        
        for(int i = 0; i<candyType.size() - 1; i++)
        {
            if(candyType[i] == candyType[i+1])
            {
                continue;
            }
            else{
                unique_element++;
            }
        }
        int set_count = unique_element + 1;
        
        int n = candyType.size() / 2;
        
        return min(set_count, n);
    }
};
// Time Complexity - O(nlogn), since we're doing sorting
// Space Complexity - O(1)

// Method - 2
class Solution {
public:
    int distributeCandies(vector<int>& candyType) {
        set<int> s;
        
        for(int &element: candyType)
        {
            s.insert(element);
        }
        
        int set_size = s.size();
        
        int vector_size = candyType.size() / 2;
        
        if(set_size < vector_size)
        {
            return set_size;
        }
        return vector_size;
    }
};

// Time Complexity - O(N), whole iteration of the list 
// Space Complexity - O(N), set is being used

File: LeetCode/Algorithms/Easy/DistributeCandies/DistributeCandies.py
class Solution:
    def distributeCandies(self, candyType: List[int]) -> int:
        set_candy = set(candyType)
        len_set_candy = len(set_candy)
        
        candy_len = len(candyType)
        
        num_candies = candy_len / 2 
        
        if(len_set_candy < num_candies):
            return int(len_set_candy)
        else:
            return int(num_candies)
# Time Complexity - O(N), since iterating over a list is O(N) and adding each element to the hast set is O(1), so the total operation is O(N).
# Space Complexity - O(N), using set
